File: frontend/src/api/billing.ts
import axiosInstance from './axiosInstance';

export interface CreditPack {
  price_id: string;
  credits: number;
  label: string;
  price_display?: string | null;
}

export interface SubscriptionStatus {
  has_subscription: boolean;
  status?: string | null;
  cancel_at_period_end: boolean;
  current_period_end?: string | null;
  stripe_subscription_id?: string | null;
  price_id?: string | null;
}

export interface BillingInvoice {
  invoice_id: string;
  status?: string | null;
  amount_paid: number;
  amount_refunded: number;
  currency?: string | null;
  created?: string | null;
  hosted_invoice_url?: string | null;
  grant_id?: string | null;
  grant_remaining: number;
  refundable: boolean;
}

export async function listCreditPacks(): Promise<{
  enabled: boolean;
  packs: CreditPack[];
}> {
  const { data } = await axiosInstance.get('/billing/credit-packs');
  return data;
}

export async function createCheckoutSession(priceId: string): Promise<{
  url: string;
  session_id: string;
}> {
  const { data } = await axiosInstance.post('/billing/checkout-session', {
    price_id: priceId,
  });
  return data;
}

export async function getSubscriptionStatus(): Promise<SubscriptionStatus> {
  const { data } = await axiosInstance.get('/billing/subscription');
  return data;
}

export async function cancelSubscription(): Promise<SubscriptionStatus> {
  const { data } = await axiosInstance.post('/billing/subscription/cancel');
  return data;
}

export async function resumeSubscription(): Promise<SubscriptionStatus> {
  const { data } = await axiosInstance.post('/billing/subscription/resume');
  return data;
}

export async function listUserInvoices(userId: string): Promise<BillingInvoice[]> {
  const { data } = await axiosInstance.get(`/billing/users/${userId}/invoices`);
  return data.invoices ?? [];
}

export async function refundUserInvoice(
  userId: string,
  invoiceId: string,
  revokeCredits: boolean
): Promise<{
  refund_id?: string | null;
  status?: string | null;
  amount?: number | null;
  currency?: string | null;
  invoice_id: string;
  revoked_grant_id?: string | null;
}> {
  const { data } = await axiosInstance.post(`/billing/users/${userId}/refunds`, {
    invoice_id: invoiceId,
    revoke_credits: revokeCredits,
  });
  return data;
}

File: backend/bebcare/schemas/billing.py
from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, Field


class CreditPackOut(BaseModel):
    price_id: str
    credits: int
    label: str
    price_display: str | None = None


class CreditPacksResponse(BaseModel):
    enabled: bool
    packs: List[CreditPackOut]


class CheckoutSessionCreate(BaseModel):
    price_id: str = Field(..., min_length=1)


class CheckoutSessionResponse(BaseModel):
    url: str
    session_id: str


class SubscriptionItemOut(BaseModel):
    stripe_subscription_id: str
    price_id: Optional[str] = None
    label: Optional[str] = None
    status: Optional[str] = None
    cancel_at_period_end: bool = False
    current_period_end: Optional[datetime] = None


class SubscriptionActionRequest(BaseModel):
    stripe_subscription_id: str = Field(..., min_length=1)


class SubscriptionStatusResponse(BaseModel):
    has_subscription: bool
    status: Optional[str] = None
    cancel_at_period_end: bool = False
    current_period_end: Optional[datetime] = None
    stripe_subscription_id: Optional[str] = None
    price_id: Optional[str] = None
    label: Optional[str] = None
    subscriptions: List[SubscriptionItemOut] = Field(default_factory=list)


class InvoiceOut(BaseModel):
    invoice_id: str
    status: Optional[str] = None
    amount_paid: int = 0
    amount_refunded: int = 0
    currency: Optional[str] = None
    created: Optional[datetime] = None
    hosted_invoice_url: Optional[str] = None
    grant_id: Optional[str] = None
    grant_remaining: int = 0
    refundable: bool = False


class InvoiceListResponse(BaseModel):
    invoices: List[InvoiceOut]


class RefundCreate(BaseModel):
    invoice_id: str = Field(..., min_length=1)
    revoke_credits: bool = False


class RefundResponse(BaseModel):
    refund_id: Optional[str] = None
    status: Optional[str] = None
    amount: Optional[int] = None
    currency: Optional[str] = None
    invoice_id: str
    revoked_grant_id: Optional[str] = None
